File: 13. Roman To Integer/13RomanToInteger.py
class Solution(object):

    def romanToInt(self, s):
             
        # Roman to Integer Mapping

        romanLettersValue = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000
        }
        
        # It's easier to solve roman numerals from the end to the beginning
        # So we reverse the string

        s = s[::-1]

        # We will store the sum of the roman numerals to return it later
        sum = 0

        #For each letter in the string we will check if it's a roman numeral
        for letter in s:
            if letter in romanLettersValue:

                # If it is a roman numeral we will add it to the num
                num = romanLettersValue[letter]

                # If 4 times the number is bigger than the sum, it means that it is those specific cases like (IV, IX, XL, XC, CD, CM)
                # So we will subtract the number from the sum

                # Example: IV = 4
                # First Interation: sum = 0, num = 5: sum = 5
                # Second Interation: sum = 5, num = 1: (as 4 * num < sum) sum = 4

                if 4 * num < sum: sum -= num
                else: sum += num

        return sum

File: 13. Roman To Integer/13RomanToInteger.c
int romanToInt(char * s){

    // num will save the charater vaule, ans the total value of the roman number
    int num, ans = 0;

    // For each character in the string, counting from the end
    for(int i = strlen(s) - 1; i >= 0; i--)
    {
        // We get the value of the character
        switch(s[i])
        {
           case 'I' : num = 1; break;
           case 'V' : num = 5; break;
           case 'X' : num = 10; break;
           case 'L' : num = 50; break;
           case 'C' : num = 100; break;
           case 'D' : num = 500; break;
           case 'M' : num = 1000; break;
        }
        
        // And for every exception like IV, IX, XL, XC, CD, CM, where the value of the character is less than the value 
        // of the next character, we need to subtract the value of the character to the answer
        if (4*num < ans)
        {
            ans -= num;
        } else {
            ans += num;
        }
    }
    
    return ans;
}